docs/reference/api/asyncresult.rst

```diff
@@ -0,0 +1,100 @@
+AsyncResult
+===========
+
+An async-aware :doc:`result` counterpart.
+
+Can be combined with asynchronous code without having to ``await`` anything right until
+the moment when you're ready to extract the final ``Result`` out of it.
+
+Can also be combined with synchronous code for convenience.
+
+.. code-block:: typescript
+
+    // T is the Ok value type, E is the Err error type
+    AsyncResult<T, E>
+
+Imports:
+
+.. code-block:: typescript
+
+    import { AsyncResult } from 'ts-results-es'
+
+Construction:
+
+You can construct it directly from ``Result<T, E>`` or ``Promise<Result<T, E>>``:
+
+.. code-block:: typescript
+
+    const result1 = new AsyncResult(Ok(1))
+    const result2 = new AsyncResult((async () => Err('bad_error')())
+
+Or you can use the :ref:`Result.toAsyncResult() <toAsyncResult>` method:
+
+.. code-block:: typescript
+
+    const result3 = Ok(1).toAsyncResult()
+
+Only the most important methods are currently implemented. The following methods can be
+added with not too much effort:
+
+* ``mapErr()``
+* ``mapOr()``
+* ``mapOrElse()``
+* ``or()``
+* ``orElse()``
+
+``andThen()``
+-------------
+
+.. code-block:: typescript
+
+    andThen<T2, E2>(
+        mapper: (val: T) => Result<T2, E2> | Promise<Result<T2, E2>> | AsyncResult<T2, E2>
+    ): AsyncResult<T | T2, E | E2>
```

Comment:
What is `E` here?

Reply:
The type of the original error (if the result is an error (type `E`) then `andThen()` keeps it and doesn't call `mapper`.